### ISSUE TITLE
Papers Notes

### ISSUE THREAD

**Issue by innaamogolonova** [innaamogolonova/PB_extension#32]
Notes from read papers. Don't forget to refer to #33 !

Template: 

-> label, purpose
**Title**
link

Relevant Concepts:
- notes 

Testing Strategy or Concrete Cases:
- notes

References to Possibly Explore:
- notes

Most Relevant Takeaways (for Our Tool):
- notes

**Comment by innaamogolonova** [comment]
-> evidence/justification of usefulness
-> general LP 
-> tool inspiration/origin 

**Projection Boxes: On-the-fly Reconfigurable Visualization for Live Programming**
https://dl.acm.org/doi/10.1145/3313831.3376494

Relevant Concepts: 
- Notion of live programming (LP) is to provide immediate feedback each time the program is modified -> need to remember this for this extension
- Essential to keep it configurable and customizable to user preferences on-the-fly 
- Compute the set of all program states that occur at every line 
- On the box, there is a dot (green, orange, or red) representing if the state is up to date or outdated -> might be something to consider 
- Projections defined under here can do to things: pick a subset of values to display and switch display between column and row order
- Specifies difference in user preferences in code-writing mode vs debugging. Users generally want less information in the former and more in the later. -> I think debugging applies more to this tool, need to keep concrete case context within debugging 
- NEED keyboard shortcuts over mouse interactions 
- On/off toggle is useful for users 
- Users have to have the ability to configure what variables are displayed and how they are displayed (on/off at minimum)
- Run-always approach is tricky to implement 
- Implementation here ran the function over a series of unit tests and extracted the values that way 


Testing Strategy or Concrete Cases: 
- Pretty small scale and on isolated test cases over simple manipulations in Python 

References to Possibly Explore: 
- Victor Visualization [19] + Alfie [1] and Seymour [9, 10] 
- Omnicode [8]
- Babylonian editor [16, 17] 
- The Yin Tang system [14] 
- Look into papers that cite this one

Most Relevant Takeaways for This Tool: 
- Customizable and configurable for the user 
- Keyboard shortcuts 
- Remember fundamental live programming and projection box concepts in this implementation 
- Look into Victor Visualization

**Comment by innaamogolonova** [comment]
-> evidence/justification of usefulness
-> general LP 
-> tool inspiration/origin 

**Omnicode: A Novice-Oriented Live ProgrammingEnvironment with Always-On Run-Time Value Visualizations**
https://dl.acm.org/doi/epdf/10.1145/3126594.3126632

Relevant Concepts:
- Displays the entire history of all run-time values 
- Does not scale but doesn't need to scale, targeted specifically at novice programmers with minimal scope
- While the user is coding, displays visualizations almost real-time as the code updates 
- Implements a simpleform of continuous testing [25] -> running code on currently selected test case 2 seconds after typing stops 
- Findings raise the possibility of Omnicode’s always-on visualizations leading to serendipitous visual discoveries and their uses as communication aids in addition to being debugging tools.
- One compromise we can pursue in the future is to still keep Omnicode’s unique “DISPLAY ALL THE VALUES!” spirit, but to also add automatically-generated suggestions of “interesting” visualizations and code regions for users to focus on. -> literally what PB Ext is 
- Omnicode 2.0 could determine interestingness based on static code analysis (e.g., variables i and always appear close together syntactically), dynamic analysiakin to Daikon [11] (e.g., i and j always take on strongly-correlated values), or by tracking user interactions within theIDE such as keyboard and mouse locations to determine whathe user likely cares about the most at any given moment. -> possible hints to guide LLM's critical point selection? 

Testing Strategy or Concrete Cases:
- hand crafted small function scenarios to implement 

References to Possibly Explore:
- Bret Victor’s LearnableProgramming essay [31]
- Tanimoto’s classiﬁcation [30] 
- Python Tutor [13]

Most Relevant Takeaways for This Tool:
- can this idea be applied to a real world context? more scalable? 
   - maybe not all runtime values but have some easy access to them 
   - i guess that makes it not this concept 
-  limitations and future directions can lead to better implementation

**Comment by innaamogolonova** [comment]
-> benchmark
-> explore more!

**BugsInPy: A Database of Existing Bugs in Python Programs toEnable Controlled Testing and Debugging Studies**
https://dl.acm.org/doi/epdf/10.1145/3368089.3417943

Relevant Concepts:
- contains 493 bugs from 17 real-world python projects 
- bugs are isolated, which is good for reproduction 
- from popular open source projects 
- encompasses lots of topics 

Most Relevant Takeaways (for Our Tool):
- dig through this repo and find possible relevant cases: https://github.com/soarsmu/BugsInPy
- possible implementation: give user a bug and see how they will fix it using the runtime values 
- needs further digging through the actual benchmarks to find relevant ones

**Comment by innaamogolonova** [comment]
-> benchmark
-> explore more!

**Tests4Py: A Benchmark for System Testing**
https://arxiv.org/pdf/2307.05147

Relevant Concepts:
- 73 bugs from real-world applications and 6 from example programs 
- provides oracle for functional correctness and supports system and unit test integration 
- built on top of BugsInPy

Most Relevant Takeaways (for Our Tool):
- dig through this repo and find possible relevant cases: https://github.com/smythi93/Tests4Py
- possibly use unit tests to guide LLM filtering or runtime value production (like in the PB paper)
- possible implementation: give user a bug and see how they will fix it using the runtime values 
- needs further digging through the actual benchmarks to find relevant ones

**Comment by innaamogolonova** [comment]
-> explore more!
-> case studies 
-> DL

**Silent bugs in deep learning frameworks: an empirical study of Keras and TensorFlow**
https://link.springer.com/article/10.1007/s10664-023-10389-6#Sec21

Relevant Concepts:
- 77 reproducible bugs and root causes
- specifically silent bugs
- in tensorflow and keras
- not sure how a LP environment might be helpful but can try
- especially interesting categories (imo): wrong parameter setting, wrong displayed message, wrong resulting shape, wrong resulting structure
- github with reproducible bugs: https://github.com/amin-nikanjam/SilentBugsInTensorFlowKeras 

Testing Strategy or Concrete Cases:
- manually analyzed and selected 77 issues 
- conducted an interview with 103 users on relevance

References to Possibly Explore:
- [On the Soundness of Silence: Investigating Silent Failures Using Fault Injection Experiments](https://ieeexplore.ieee.org/document/6821096)
- [A comprehensive study of real-world numerical bug characteristics](https://ieeexplore.ieee.org/document/8115662)
- [An empirical study on TensorFlow program bugs](https://dl.acm.org/doi/abs/10.1145/3213846.3213866)

Most Relevant Takeaways (for Our Tool):
- could live programming help in this case?

**Comment by innaamogolonova** [comment]
-> DL 
-> justification

**A Comprehensive Study on Deep Learning Bug Characteristics**
https://dl.acm.org/doi/epdf/10.1145/3338906.3338955

Relevant Concepts:
- comprehensive study of bug patterns in 5 popular DL libraries (Tensorflow and Keras incl) 
- possibly relevant domains of bugs: API bug, logic bugs 
- possible useful root causes: incorrect model parameter (IPS), structural inefficiency (do not cause crashes), absence of type checking 
- can't find their data/gh  

Most Relevant Takeaways (for Our Tool):
- some justification for looking into DL and (possibly) type checking bugs in python

**Comment by innaamogolonova** [comment]
-> DL
-> explore more!

**An Empirical Study on TensorFlow Program Bugs**
https://dl.acm.org/doi/epdf/10.1145/3213846.3213866

Relevant Concepts:
- bugs from deep learning applications built on top of tensorflow 
- dataset available at https://github.com/ForeverZyh/TensorFlow-Program-Bugs
- IPS, API misuse, API change, could all be interesting to explore 
- referenced by a lot of other papers
- need to read more carefully and dig around the dataset

**Comment by innaamogolonova** [comment]
-> justification 
-> definitions/concepts 

**Debugging into Existence with Program Synthesis**
https://ieeexplore.ieee.org/stamp/stamp.jsp?tp=&arnumber=11303493

Relevant Concepts:
- exploratory programming, debugging into existence, live execution
- DeSYnt: debugger extensions to leverage program synthesis during Debugging into Existence 
- DESYNT combines debugging by live execution past a predetermined hole with a PBE synthesizer, that can suggest hole completions to the programmer
- very program synthesis(y) 
- can draw inspiration/design ideas from here 
- for partial execution on code, not exactly our focus but could be useful 
- tool proved useful for API discovery where functionality is simple but the implementation is difficult or uses an unfamiliar API
- cites lots of papers by n. polikarpova 

Testing Strategy or Concrete Cases:
- 3 custom test tasks (new feature 1 file ~70 LOC, bug fixing 24 files ~1900 LOC, reverse engineering 2 files ~141 LOC)
- did not find specific motivation for the chosen tests

References to Possibly Explore:
- 44, 41, 2, 4, 14
- 5: hypothesizer - uses a database of hypotheses and tries to match them with demonstrated faulty behavior 
- 12, 23: about api discovery 

Most Relevant Takeaways (for Our Tool):
- the general area of work of this paper seems really relevant, even if the tool itself is a little bit adjacent to PB 
- could be a good reference for relevant terms/concepts; look in the related work section

**Comment by innaamogolonova** [comment]
-> case study
-> "benchmark"

**Blindspots in Python and Java APIs Result in Vulnerable Code**
https://dl.acm.org/doi/epdf/10.1145/3571850

Relevant Concepts:
- APIs with blindspots are a serious risk in developer's understanding of programs 
- need to read further into the study (how they conducted it)
- provides a very concise and clear dataset of Python puzzles that use APIs with blindspots, replicating real world uses and problems in a digestible and testable way 
   - limitation: are these too concise? 
- data package available here: https://osf.io/ahpfv/overview?view_only=37978cb1e67941d5b1e572f2fba982c9

Most Relevant Takeaways (for Our Tool):
- can these blindspots be revealed and avoided with LP/PB?

**Comment by innaamogolonova** [comment]
-> general 

**TraceDiff: Debugging Unexpected Code Behavior Using Trace Divergences**
https://ieeexplore.ieee.org/stamp/stamp.jsp?tp=&arnumber=8103457

Relevant Concepts:
- automatic hint delivery in teaching/tutoring domain 
- pretty elementary/simple tasks 
- maybe useful to explore implementation
- might be useful to come back to if pivoting to cs tutoring domain (unlikely)

References to Possibly Explore:
- 17, 13: program visualization tools that execute the program, save snapshot of internal states at each execution step and produce a visual representation of runtime states (beginner programmer specific)
- 21, 22: overview of execution behavior and let a user find the cause of the bug through interactive qa or retroactive logging 
- 23, 24, 25:  important divergent runtime behavior can be detected and highlighted in the domain of web application debugging 

Most Relevant Takeaways (for Our Tool):
- provides a lot of useful references 
- look into design for implementation

**Comment by innaamogolonova** [comment]
Random (potentially) relevant papers: 
- https://ieeexplore.ieee.org/document/9833127
   - analysis of front-end development challenges from stackoverflow posts 
   - might be useful if designing your own web request task 

- https://ieeexplore.ieee.org/document/11303460 
   - similar idea to PB extension: automatic insertion of print statements 
   - interesting design: ppants asked to use the tool in their regular daily programming and then interviewed about it 

- https://dl.acm.org/doi/10.1145/2556288.2557409 
   - similarly, asked ppants to use tool in day-to-day work 

- [Online python tutor: embeddable web-based program visualization for cs education](https://dl.acm.org/doi/10.1145/2445196.2445368)
   - + look at other Philip Guo papers, seems to have a lot of work in this area

**Comment by innaamogolonova** [comment]
-> motivation for hypothesis feature 

**Hypothesizer: A Hypothesis-Based Debugger to Find and Test Debugging Hypotheses**
https://dl.acm.org/doi/pdf/10.1145/3586183.3606781

Relevant Concepts:
- could be relevant for implementing the user entered hypothesis to help aid in debugging 

Testing Strategy or Concrete Cases:
- inserted bugs into open source web apps, which could be cool 

Most Relevant Takeaways (for Our Tool):
- justifying the hypothesis feature

**Comment by innaamogolonova** [comment]
[An Empirical Study on Dynamic Typing Related Practices inPython Systems](https://dl.acm.org/doi/epdf/10.1145/3387904.3389253)
- maybe useful if looking at dynamic typing related bugs 
- deals with very large OOS projects that would be difficult to scale to

**Comment by blakenewhouse** [comment]
-> code comprehension
-> justification(?)
**How Do Software Engineers Understand Code Changes? - An Exploratory Study in Industry** 
https://taoxie.cs.illinois.edu/publications/fse12-study.pdf
suggested (by Nadia i think?) at Apr 23 mtg

Relevant Concepts:

- Code comprehension ideas, from industry
- Justifies code comp
- really interesting ranking on "information needs' (figure 5a), shows what programmers most want when reviewing code changes (top 3 are: Rational, Correctness, Risk)

Testing Strategy or Concrete Cases:

- N/A but paper is good justification for controlled studies vs industry

References to Possibly Explore:

- 4, 9: related to code comprehension//live coding
- 32: developer mental models

Most Relevant Takeaways (for Our Tool):

- Most important bit of information, but also hardest to acquire, is "does this change break any code elsewhere?"
    - This correlates with "how does this change alter the program’s dynamic behavior?”, which is maybe what we can provide?
- live tool allows for instant feedback on a change
